### DISCUSSION TITLE
May I ask which routing algorithm does organicmaps use?

### DISCUSSION THREAD

**Discussion by BluEe33** [organicmaps/organicmaps#4756] [Q&A] [answered]
As an offline navigation app, does it use the A* algorithm?

**Comment by biodranik** [comment, marked as answer]
OM uses bidirectional A*

**Reply by BluEe33** [reply]
Thanks for answering！ Is the function of this bidirectional A* algorithm in framework.cpp? Or is it another file?

**Reply by biodranik** [reply]
Check the routing library. Why do you need it? What is your goal/need?

**Reply by BluEe33** [reply]
I am a college student, and my graduation project is to do optimization on organicmaps. The most important thing is the route planning algorithm. Thank you！

**Reply by biodranik** [reply]
That sounds great! What kind of optimization do you plan to work on?

**Reply by BluEe33** [reply]
Sometimes the path chosen by the navigation is not the shortest path, I need to figure out why and optimize it.

**Reply by ghost** [reply]
As far as I know it always chooses the fastest route only but you got my attention as I asked for shortest route as an option some time ago.

**Reply by davidcollini** [reply]
I always guessed that it didn’t always choose the shortest path because it doesn’t always have a speed limit for every road, but I never looked into it ;P

**Reply by biodranik** [reply]
@BluEe33 please check the related OpenStreetMap data and the speed model. Currently, the _fastest_ path is chosen, not the _shortest_ one. It would be great if you can help us implement alternate routes, including the fastest one. Somewhere was an issue that listed other proposed alternate criteria.

**Reply by BluEe33** [reply]
Sorry, I checked the routing library, but didn't find a code that implements the bidirectional A* algorithm concretely. Can you tell me in which file the implementation function of this algorithm can be found? Thank you so much!

**Reply by biodranik** [reply]
Parts of the algorithm are implemented in different files in the routing directory. Please find an entry point in router.hpp

**Reply by BluEe33** [reply]
Thank you very much for your answer! Where is the file location of the heuristic function?
